### ISSUE TITLE
Problem with OCR with Linux

### ISSUE THREAD

**Issue by moonchild-it** [SubtitleEdit/subtitleedit#1989]
I am trying to use Subtitle Edit 3.4.13, under Ubuntu 16.04.

When I convert subs via OCR, accented letters are considered as error, 
I suppose due a bad encoding setting.

I launch Subtitle Edit with mono , version 4.2.1

I have mono version 4.2.1
Hunspell version 1.3.3
Tesseract 3.04.01

This is my localization:

LANG=it_IT.UTF-8
LANGUAGE=it_IT:en_GB:en

I've found a very ugly and dirty workaraund.
I renamed /usr/bin/tesseract to /usr/bin/tesseract.bin
and created this script called  /usr/bin/tesseract

```
#!/bin/sh

/usr/bin/tesseract.bin $@
/usr/bin/iconv -f UTF-8 -t ISO-8859-2 $2.hocr > /tmp/dep
mv -f /tmp/dep $2.hocr
```

This brokes the funcionality of tesseract under linux obvisuly.

Sorry for the bad english.

**Comment by niksedk** [comment]
Do you know if the .hocr file is always utf-8?

**Comment by moonchild-it** [comment]
tesseract produces always utf-8 because is my system language.

The problem is that ISO-8859-2 is not the right encoding, very close but
not the right one for nhunspell. Some chars like é and ì are different.

I am trying to find this encoding at this moment because i need almost 
a workaraund.

I tried to save all png generated to a directory and used linux tesseract 
from shell, it produces utf-8 and linux hunspell command works fine.

Thx for the reply.

**Comment by moonchild-it** [comment]
I did some investigations.

I saved all output of tesseract launched from subtitle_edit in a directory in text
format. I concatenated all file in a single one and launched a chardet to the resulting
file, in all case the output is in utf-8 format.

The problem is it with Hhunspell, I downloaded the source code from [http://www.crawler-lib.net/nhunspell](http://www.crawler-lib.net/nhunspell). I don't know anyting of C# and .NET technologies but I understood that this .dll can handle different encoding format, and convert in utf-8 to send to libhunspell.so

The problem is why considers such input not utf-8

I tried to compile it in my linux machine with enabled debugging informations using the xbuild command but received only a lot of errors that I don't know how to fix it.

Today i slept all the day but now I want to write in the nHunspell forum.

**Comment by moonchild-it** [comment]
[http://www.crawler-lib.net/boards/topic/65/encoding-problem](http://www.crawler-lib.net/boards/topic/65/encoding-problem)

**Comment by niksedk** [comment]
I think it's .net that somehow says the file is not utf-8...
I've tried to force utf-8 here: https://github.com/SubtitleEdit/subtitleedit/releases/download/3.4.13/SubtitleEditBeta.zip
Commit is here: https://github.com/SubtitleEdit/subtitleedit/commit/f9b901740d4f5491293f5e2ade98e26a3870bef8

It is possible to detect if there are any non-utf-8 sequences but that seems overkill... I hope.

**Comment by moonchild-it** [comment]
I tried this patch but the problem persists, here a screenshot:
[http://www.moonchild.it/sub_edit_problem.png](http://www.moonchild.it/sub_edit_problem.png)

A strange thing is that the software checks for installed OCR tessdata under own directory Tesseract/tessdata, but really use the one installed under /usr/share/tesseract-ocr/tessdata/ (becouse don't uses the Tesseract.exe, but the linux tessaract installation).

Anyway I completely purged and reinstalled all ubuntu packages in order to do these tests, to avoid misconfigurations. For this reason sounds me strange that no1 have the same problem, ubuntu is a very popular distribution, and I have not customized nothing regarding tesseract, hunspell or mono.

Anyway thx for your support...

**Comment by moonchild-it** [comment]
I fixed the problem, I know how to fix it, I did some try and should works.

What u need the encoding of the word (for now we suppose UTF-8, but in general you can
retrieve it from environment variable LANG, the part after the dot). In my case

LANG=it_IT.UTF-8

U need to target encoding, this is written in the .aiff file in the line after the word SET:

```
moonchild@darkmoon:~/bin/SubtitleEdit$ grep SET ./Dictionaries/it_IT.aff
SET ISO8859-15
```

Use an equivalent of C of iconv and encode the input word from input encoding to the
target encoding, in my case in C is:

```
i_fd = iconv_open("ISO-8859-15", "UTF-8");
iconv(i_fd, &tempw2, &a, &tempw2, &b);
iconv_close(i_fd);
```

The call the ::spell method of Hunspell with the encoded word, without specifing
any encoding.

I tried in C and it works, I translated a Milla Jovovich movie without an error. :)

I don't know how implement in C#, but in any case I had successifully compiled
your code under Linux, but I had to rename a lot of file beacouse Linux fs is
case sensitive, and a lot if filenames are different from the .csproj

Thx for the support and patience.

**Comment by niksedk** [comment]
You should probably look at this file:
https://github.com/SubtitleEdit/subtitleedit/blob/master/src/Logic/SpellCheck/LinuxHunspell.cs

Is the Hunspell dictionary not utf-8?
Perhaps you could try one of these: https://github.com/titoBouzout/Dictionaries

**Comment by moonchild-it** [comment]
A hunspell dictionary seems that can have a own encoding that is it declared in the .aff file using the keyword SET. The italian dictionary shipped with hunspell is ISO-8859-15, I don't know the others.

Anyway the dictionaries that you linked now (https://github.com/titoBouzout/Dictionaries) are all converted in UTF-8, and these works on my system which is UTF-8 without any sort of conversion and, in my case, solve the issue.

The solution that I have suggested solve the problem in general, also with the standard dictionaries.

This night I modified the libhunspell.so, written in C, converting all input (that I logged in a file and verified that was UTF-8) in ISO-8859-15. In this way I have too solved the issue with the official hunspell libraries.

I suggested this solution because i logged the output of Linux hunspell command and it does something similar to the procedure that I have suggested.
Really the hunspell command is a bit more sophisticated, bcs instantiates a class AffixMgr of the libraries that parse the .aff file and returns the correct encoding set, in this case ISO-8859-15.
In other the hunshell command manage UTF-16 in different way that I not studied bcs felt to sleep :)

Anyway personally I solved my problem or with my hacked libhunspell or with your convereted UTF-8 dictionaries.

If you want extends the solution, in more general ways, read the .aiff file and where you fine the "SET" word there is declared the encoding of the .dic file. Just implements a conversion from the encoding set in the LANG environment file.

This solution should works in all situations, I hope almost.

Unfortunately my english is not very good, is a language that my brain refuse to learn. I would want speak better also because i was all night studying, modifying and tracing code of hunspell and of your great software.

I am happy because i solved the problem, and I learned a lot of things about worlds like .NET just now for me totally unknown.

Thanks another time for your support.

Best regards.

**Comment by moonchild-it** [comment]
Hello,
I discovered now another interesting thins.

The problem with OCR, using the UTF-8 dictionaries, with my configuration, does'nt exists anymore, but, if I start the spell checker from the Menu (in the main GUI then, outside the OCR panel), the problem persists.

I supposed that in this case it uses another path for dictionaries, even if, if I open the dictionaries directories from the GUI, it shows the correct path. I tried the to link the default nuspell dir (/usr/share/hunspell) to the SubEdit dir, nothing changed.

It's very strange this different behavior from the same software, with the same functionality,  in two different instances.

For now I am checking existing subtitle with Aegisub, but I'd like use Subtitle Edit in order to have the same dictionary.

In this moment due to personal problems I can't investigate anymore, and my ripping activities must be postponed.

Best regards.
